### DISCUSSION TITLE
Pybricks for build hat? In python on pi?

### DISCUSSION THREAD

**Discussion by gregorianrants** [pybricks/support#1960] [Ideas]
Is there any interest in making a Pybricks for the buildhat?

What about just doing it in python on the raspberry pi?

I have read in other places that there are issues with the rate sensors can be read.  There is a new command that allows changing the rate data is output at.

Using selrate(10) with the motors causes data to be output every 10ms

Not sure how it works with other sensors as I don’t have any. I just use gpio sensors.

I have implemented my own attempt at this in python you can find it here
https://github.com/gregorianrants/buildhat-alternative

It only does speed control so far. I haven’t looked at position control. I have played about with stall detection and had success but it hasn’t been pushed to the repo yet, as the code need polish.

I have previously mentioned doing this in another question, but the previous version was in nodejs and was not a standalone library.

The current version is fully in python it loads the firmware its self and can be installed in other projects.

To see it in action you can have a look at this project.
https://github.com/gregorianrants/composed-robot

**Comment by laurensvalk** [comment]
Thanks for your note. See https://github.com/pybricks/support/issues/812 for some previous discussion.

It would be quite doable to port the Pybricks motors and sensor drivers to the Raspberry Pi Build Hat. However, for some reason loading other firmware seems to require a signing key that we do not have. See also https://github.com/RaspberryPiFoundation/python-build-hat/issues/135

Doing it in Python and interfacing with the existing Build Hat firmware isn't going to give a very good experience, since it would have all the same problems and limitations as the current implementation.

**Reply by gregorianrants** [reply]
Hi laurence

Thanks for your feedback.  And all the great work you guys do.  I also have your book which is great.

what about if the limitations were surmountable, i think i may have demonstrated that some of them can be overcome.  though my project is not pybricks quality it may have showed that it is worth looking into again.

what i have overcame:

- i have managed to read the motor sensors with a rate of 100hz, due to an update the firmware and a command the allows changing the rate, a lot faster than in the linked to video.
- I have implemented pid on the pi rather than using the built in pid on the build hat. I read the tachos on the motors using the increased sensor rate and update the power via setting pwm directly.  this is not how the current library does it.  I think the pid implementation on the build hat is a the source of a lot of the problems.
there is a graph, on the repo linked to above of the speed i get when doing pid.  It could probably be tuned some more as i worked on other things, once i got it good enough, to start implementing behaviours. not going to be as good as pybricks though but maybe still decent enough
- this has allowed me to overcome the motors dont start at once problem https://github.com/RaspberryPiFoundation/python-build-hat/issues/152
- I have managed to implement stall detection though it is not in the repo yet.

all of this has allowed me to implement a fairly complex self driving robot that knows where it it and never crashes.

maybe its still not going to be good enough, but i think it can be better than it previously seemed.

**Reply by laurensvalk** [reply]
Glad to hear you were able to achieve improvements via Python already!

We would still be happy to consider a firmware solution if the Raspberry Pi teams reaches out to us about what to do with the signing key. (We don't actually need any source code if that is their issue.)

**Reply by gregorianrants** [reply]
Lets hope they do that one day, would be great to see.

**Reply by AndyA** [reply]
Whoop! https://www.raspberrypi.com/news/build-hat-firmware-now-fully-open-source/

**Reply by laurensvalk** [reply]
See also https://github.com/pybricks/support/issues/812

**Reply by joubarc** [reply]
Considering the Build Hat seems to host a RP2040 which is micropython-capable, and that power come form a dedicated plug on the hat itself... I was also wondering (assuming PyBricks ever gets ported to it), if it could work autonomously - that is, not plugged on a Pi Board at all. Of course you'd need to be able to communicate with it (I doubt the Build Hat itself has bluetooth) somehow, but maybe it would be possible to do that while connected; and once done, unplug it and have it run autonomously...
In any case, even with a cheap Pi underneath, the fact that it's a not-too-expensive solution without batteries required can be very interesting for event where you need your model to run continuously (*couch* GBC *cough*). I'm already using the Build Hat Power supply to feed a Technic Hub anyway.

**Reply by dlech** [reply]
This is certainly possible. Perhaps the simplest solution would be to use an FTDI UART chip that also has GPIOs. This would provide a connection to the serial port for uploading firmware/user programs and a GPIO is needed for reset.

Also, FYI, we do have a working build of Pybricks for the Build HAT now. The I/O port driver isn't implemented yet, so it doesn't do much, but it is a start.
